### PR TITLE
Location href in about blank src iframe

### PR DESCRIPTION
**Pull request by SimonMueller**
Fixes https://github.com/jsdom/jsdom/issues/3107

### CODE REVIEW

**Review by domenic — APPROVED** [review]
Thanks very much! I cleaned up the tests a bit. Interestingly while Firefox passes them all, Chrome appears to have a bug where it generates `about:blank#blocked` for the invalid URL case.

Let's see if we can get the CI passing (now that it's green on master) and then I'll merge this.

**Comment by domenic** [comment]
Hmm, it looks like the code changes you made break several other tests:

- html/semantics/document-metadata/the-base-element/base_about_blank.html
- html/semantics/document-metadata/the-base-element/base_srcdoc.html
- html/semantics/scripting-1/the-script-element/execution-timing/101.html (timeout)
- webstorage/document-domain.html (timeout)
- webstorage/event_basic.html
- webstorage/event_case_sensitive.html
- webstorage/event_no_duplicates.html
- xhr/open-url-about-blank-window.htm

Maybe one thing to consider is that such iframes should have an origin equal to their parent, even if they have a URL of `about:blank`.

Let me know if you can get those tests passing, and then we can consider merging this.

**Review by domenic — CHANGES_REQUESTED** [review]
Needs to pass the existing test suite.

**Comment by SimonMueller** [comment]
@domenic Thanks for having a look at it. When i merged upstream i saw that those tests are failing but I wasn't sure why and couldn't make them pass right away. Currently I am really busy at work but I will have a look at it as soon as I have time. It looks like a side effect from using iframes in the tests...

If you have another hint of where i could start searching in the code i would really appreciate it.

**Comment by SimonMueller** [comment]
@domenic ok nevermind i just couldn't leave it like that. the tests are  green now. 🚀 Your hint with the origin was pointing me in the right direction...

The only thing i wasn't able to figure out was why `document._URL` is undefined in `lib/jsdom/living/helpers/document-base-url.js` at this point.

File: lib/jsdom/living/helpers/document-base-url.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ exports.fallbackBaseURL = document => {
     return exports.documentBaseURL(document._defaultView._parent._document);
   }
 
-  return document._URL;
+  return whatwgURL.parseURL(document.URL);
```
Comment (domenic) [inline lib/jsdom/living/helpers/document-base-url.js:31]: Is this change necessary?

**Review by domenic — COMMENTED** [review]
Thanks, but it looks like this no longer merges with the master branch? Can you get it rebased so there are no conflicts?

File: lib/jsdom/living/nodes/Document-impl.js [hunk 2]
```diff
@@ -158,7 +158,9 @@ class DocumentImpl extends NodeImpl {
     }
 
     this._URL = parsed;
-    this._origin = whatwgURL.serializeURLOrigin(parsed);
+    this._origin = whatwgURL.serializeURL(parsed) === "about:blank" && privateData.options.referrer ?
```
Comment (domenic) [inline lib/jsdom/living/nodes/Document-impl.js:161]: Instead of using referrer, which is only supposed to impact document.referrer, can you just add a new origin option?

That would also avoid all this parsing and serializing. We could just pass `origin: parentDoc._origin`.

Comment (domenic) [inline lib/jsdom/living/nodes/Document-impl.js:161]: There's no need to re-serialize. Just use `urlOption`.

Comment (SimonMueller) [reply lib/jsdom/living/nodes/Document-impl.js]: good point. done.

File: lib/jsdom/living/helpers/document-base-url.js (hunk 1 above)
Comment (SimonMueller) [reply lib/jsdom/living/helpers/document-base-url.js:31]: yes it is required to pass the tests for the base element.

i debugged this and because of my change in the iframe implementation the code now reaches  `return exports.documentBaseURL(document._defaultView._parent._document);` on line 28 and as far as i can tell we only have the public api `document.URL` to work with if we access the parent document this way. thats why i cannot access `document._URL` anymore and i think this code cannot work in other scenarios aswell because of this.

if you know a better way to do this then i would be interested to see it :)

**Comment by domenic** [comment]
Hmm seeing repeated failures on v14 with the cloneNode test. I might have somehow broken it on a previous commit... I'll debug locally before merging this.

**Comment by domenic** [comment]
I can't reproduce locally; the test doesn't even take that long. Hmm. I'll just re-run CI I guess...
